docs/source/advanced/contrib_thunder.rst

```diff
@@ -0,0 +1,468 @@
+The sections below are meant to be read **sequentially**.
+
+~~~~~~~~~~~~~~~~~~
+Adding a primitive
+~~~~~~~~~~~~~~~~~~
+A good example of adding a primitive operation to the ``torch`` executor is
+the `PR #136 <https://github.com/Lightning-AI/lightning-thunder/pull/136>`_
+which adds support for `torch.Tensor.unfold <https://pytorch.org/docs/stable/generated/torch.Tensor.unfold.html>`_.
+
+Let's outline some of its key parts.
+
+The following added (emphasized) lines to
+`thunder/core/prims.py <https://github.com/Lightning-AI/lightning-thunder/blob/main/thunder/core/prims.py>`_
+
+.. code-block:: python
+   :emphasize-lines: 5
+   :lineno-start: 152
+
+   SLICE = auto()
+   SQUEEZE = auto()
+   TRANSPOSE = auto()
+   UNFOLD = auto()
+   VIEW = auto()
+   # Memory layout prims (Experimental)
+   STRIDE_ORDER = auto()
+
+.. code-block:: python
+   :emphasize-lines: 1-18
+   :lineno-start: 3082
+
+   def unfold_meta(a: TensorProxy, /, dim: int, size: int, step: int) -> TensorProxy:
+       dim = utils.canonicalize_dim(a.ndim, dim)
+       max_size = 1 if a.ndim == 0 else a.shape[dim]
+
+       utils.check(
+       size <= max_size, lambda: f"Maximum size for tensor at dimension {dim} is {max_size} but size is {size}"
+       )
+       utils.check(size >= 0, lambda: f"Size is {size} but must be >= 0")
+       utils.check(step > 0, lambda: f"Step is {step} but must be > 0")
+
+       shape = list(a.shape)
+       shape.append(size)
+       shape[dim] = (shape[dim] - size) // step + 1
+
+       return TensorProxy(like=a, shape=shape)
+
+
+    unfold = make_prim(PrimIDs.UNFOLD, "unfold", meta=unfold_meta, tags=(OpTags.SHAPE_OP,))
+
+register a primitive symbol ``unfold`` using ``make_prim`` with ``id=PrimIDs.UNFOLD``,
+``name=unfold``, and ``meta=unfold_meta``. One can see that ``unfold_meta`` follows the signature
+of the underlying ``torch.Tensor.unfold`` operation
+(so that the primitive is directly modeled after the PyTorch operation)
+with the only exception of expecting a ``TensorProxy``
+and not the ``torch.Tensor`` as its input. The rest of the function checks the inputs and returns
+a ``TensorProxy`` of the appropriate shape. ``like=a`` means that the output will inherit the meta-data
+like ``device`` and ``dtype`` from ``a``. The primitive is also tagged with ``tags=(OpTags.SHAPE_OP,)``,
+and, therefore, is associated with shape-based operations.
+We use tags to additionally group operations for group-specific operation optimizations inside *Thunder*.
+
+Additionally, the `clang <https://github.com/Lightning-AI/lightning-thunder/blob/main/thunder/clang/__init__.py>`_ gets updated
+
+.. code-block:: python
+   :emphasize-lines: 1-3
+   :lineno-start: 1119
+
+   @clangop()
+   def unfold(a: TensorProxy, /, dim: int, size: int, step: int) -> TensorProxy:
+       return prims.unfold(a, dim, size, step)
+
+And while this symbol directly calls into the primitive and is very simple, a more involved
+example of ``clang.slice_in_dim``
```

Comment:
This is interesting, but it distracts away from the unfold example. What if it's moved to the clang theory section?